File: internal/api/bundles_test.go
package api_test

import (
	"net/http"
	"testing"
)

func starterBundleBody() map[string]any {
	return map[string]any{
		"contract_version": "composition-bundle/v1",
		"name":             "starter-site",
		"theme":            "starter",
		"pages": map[string]any{
			"home": map[string]any{
				"contract_version": "layout-composition/v1",
				"regions": map[string]any{
					"main": map[string]any{
						"blocks": []any{map[string]any{"type": "heading", "props": map[string]any{"text": "Hi"}}},
					},
				},
			},
		},
		"sample_content": []any{
			map[string]any{"type": "article", "data": map[string]any{"title": "Hello world"}},
		},
		"manifest": map[string]any{
			"requires_contract": "layout-composition/v1",
			"blocks":            []any{"heading"},
			"slots":             []any{"main"},
		},
	}
}

func TestBundlesListIs404WhenNotConfigured(t *testing.T) {
	h := testServer(t)
	rec := do(t, h, http.MethodGet, "/api/v0/bundles", nil)
	if rec.Code != http.StatusNotFound {
		t.Fatalf("GET /bundles without WithPresets = %d, want 404", rec.Code)
	}
}

func TestBundleCreateThenGetRoundTrips(t *testing.T) {
	h, cookie := testServerWithPresets(t)
	rec := doWithCookieBody(t, h, http.MethodPost, "/api/v0/bundles", cookie, starterBundleBody())
	if rec.Code != http.StatusCreated {
		t.Fatalf("POST bundle = %d, body %s", rec.Code, rec.Body.String())
	}
	created := decode(t, rec)
	id, _ := created["id"].(string)
	if id == "" {
		t.Fatalf("created bundle has no id: %v", created)
	}

	rec = do(t, h, http.MethodGet, "/api/v0/bundles/"+id, nil)
	if rec.Code != http.StatusOK {
		t.Fatalf("GET bundle = %d, body %s", rec.Code, rec.Body.String())
	}
	if decode(t, rec)["name"] != "starter-site" {
		t.Fatalf("name mismatch: %v", decode(t, rec))
	}
}

func TestBundleImportMergesPagesAndCreatesSampleContent(t *testing.T) {
	h, cookie := testServerWithPresets(t)
	rec := doWithCookieBody(t, h, http.MethodPost, "/api/v0/bundles", cookie, starterBundleBody())
	id := decode(t, rec)["id"].(string)

	rec = doWithCookieBody(t, h, http.MethodPost, "/api/v0/bundles/"+id+"/import", cookie, map[string]any{
		"theme_regions": []any{"main"},
	})
	if rec.Code != http.StatusOK {
		t.Fatalf("POST bundle import = %d, body %s", rec.Code, rec.Body.String())
	}
	result := decode(t, rec)
	compat, _ := result["compat"].(map[string]any)
	if compat["compatible"] != true {
		t.Fatalf("import compat = %v, want compatible, full body %v", compat, result)
	}
	pages, _ := result["imported_pages"].([]any)
	if len(pages) != 1 || pages[0] != "home" {
		t.Fatalf("imported_pages = %v, want [home]", pages)
	}
	created, _ := result["created_content"].([]any)
	if len(created) != 1 {
		t.Fatalf("created_content = %v, want one item", created)
	}

	rec = do(t, h, http.MethodGet, "/api/v0/layouts/home", nil)
	if rec.Code != http.StatusOK {
		t.Fatalf("GET merged layout = %d, body %s", rec.Code, rec.Body.String())
	}
}

func TestBundleImportRequiresAdmin(t *testing.T) {
	h, cookie := testServerWithPresets(t)
	rec := doWithCookieBody(t, h, http.MethodPost, "/api/v0/bundles", cookie, starterBundleBody())
	id := decode(t, rec)["id"].(string)

	rec = do(t, h, http.MethodPost, "/api/v0/bundles/"+id+"/import", map[string]any{})
	if rec.Code != http.StatusUnauthorized {
		t.Fatalf("anonymous bundle import = %d, want 401", rec.Code)
	}
}

File: sdk-js/test/bundles.test.ts
import { describe, expect, it } from "vitest";
import { GlyphuxClient } from "../src/client.js";
import { GlyphuxApiError } from "../src/errors.js";
import type { CompositionBundle } from "../src/types.js";
import { testEnv } from "./testenv.js";

async function adminClient(): Promise<GlyphuxClient> {
  const env = testEnv();
  const client = new GlyphuxClient({ baseUrl: env.baseUrl });
  await client.auth.login(env.adminEmail, env.adminPassword);
  return client;
}

function starterSite(homeRoute: string): CompositionBundle {
  return {
    contract_version: "composition-bundle/v1",
    name: `starter-site-${Date.now()}`,
    theme: "starter",
    pages: {
      [homeRoute]: {
        contract_version: "layout-composition/v1",
        regions: { main: { blocks: [{ type: "heading", props: { text: "Hi" } }] } },
      },
    },
    sample_content: [{ type: "post", data: { title: "Hello world" } }],
    manifest: {
      requires_contract: "layout-composition/v1",
      blocks: ["heading"],
      slots: ["main"],
    },
  };
}

describe("bundles", () => {
  it("save() then get() round-trips the bundle", async () => {
    const client = await adminClient();
    const saved = await client.bundles.save(starterSite(`home-${Date.now()}`));
    expect(saved.id).toBeTruthy();

    const loaded = await client.bundles.get(saved.id);
    expect(loaded.name).toBe(saved.name);
  });

  it("import() merges pages and creates sample content when compatible", async () => {
    const client = await adminClient();
    const route = `bundle-home-${Date.now()}`;
    const saved = await client.bundles.save(starterSite(route));

    const result = await client.bundles.import(saved.id, ["main"]);
    expect(result.compat.compatible).toBe(true);
    expect(result.imported_pages).toContain(route);
    expect(result.created_content?.length).toBe(1);

    const layout = await client.layouts.get(route);
    expect(layout.regions.main.blocks[0].type).toBe("heading");
  });

  it("import() declines to merge or create anything when incompatible", async () => {
    const client = await adminClient();
    const route = `bundle-decline-${Date.now()}`;
    const saved = await client.bundles.save(starterSite(route));

    const result = await client.bundles.import(saved.id, ["header", "footer"]);
    expect(result.compat.compatible).toBe(false);
    expect(result.imported_pages ?? []).toHaveLength(0);

    let caught: unknown;
    try {
      await client.layouts.get(route);
    } catch (err) {
      caught = err;
    }
    expect(caught).toBeInstanceOf(GlyphuxApiError);
    expect((caught as GlyphuxApiError).status).toBe(404);
  });

  it("save() without authentication throws a typed 401 error", async () => {
    const env = testEnv();
    const anon = new GlyphuxClient({ baseUrl: env.baseUrl });

    let caught: unknown;
    try {
      await anon.bundles.save(starterSite(`anon-${Date.now()}`));
    } catch (err) {
      caught = err;
    }
    expect(caught).toBeInstanceOf(GlyphuxApiError);
    expect((caught as GlyphuxApiError).status).toBe(401);
  });
});
